src/libpmemobj/tx.c

```diff
@@ -1116,6 +1185,95 @@ tx_realloc_common(struct tx *tx, PMEMoid oid, size_t size, uint64_t type_num,
 }
 
 /*
+ * tx_ctx_init -- initializes the transaction context
+ */
+void
+tx_ctx_init(void)
+{
+	LOG(3, NULL);
+	int result = os_once(&tx_ctx_once, tx_ctx_create_key);
+	if (result) {
+		errno = result;
+		FATAL("!pthread_once");
+	}
+
+	/*
+	 * Workaround Helgrind's bug:
+	 * https://bugs.kde.org/show_bug.cgi?id=337735
+	 */
+	VALGRIND_ANNOTATE_HAPPENS_AFTER(&tx_ctx_once);
+}
+
+/*
+ * tx_ctx_fini -- destroys the transaction context
+ */
+void
+tx_ctx_fini(void)
+{
+	LOG(3, NULL);
+	void *ctx = os_tls_get(tx_ctx_key);
```

Comment:
Is this enough?
Are you guaranteed that `tx_ctx_fini` is called from the main thread (the one that didn't call the key destructor)?
What happens when someone calls `dlclose` in a non-main thread on libpmemobj.so and the library is unloaded?